### PR TITLE
Fix gradients

### LINKED ISSUE #20790: paddle 计算double grad 报错
## 本地环境：
cuda 9, cudnn7, python3.6 
## 问题描述，希望实现的功能是：
y = f(x)
dydx = fluid.gradient(y, x)
dydxdy = fluid.gradient(dydx*x,  y)
## 代码：
```
import paddle.fluid as fluid
import numpy as np

def test_grad_grad():
    """validation the validation of mul double grad"""
    train_program = fluid.Program()
    start_program = fluid.Program()
    place = fluid.CUDAPlace(0)
    #place = fluid.CPUPlace()
    with fluid.program_guard(train_program, start_program):
        rng = np.random.RandomState(0)
        rng = np.random
        inp_ = rng.uniform(-1, 1, [32, 24]).astype('float32')
        w1_ = rng.uniform(-1, 1, [24, 50]).astype('float32')
        w2_ = rng.uniform(-1, 1, [50, 14]).astype('float32')
        yg_ = rng.uniform(-1, 1, [2, ]).astype('float32')

        inp = fluid.layers.data('inp', [32, 24], append_batch_size=False)
        w1 = fluid.layers.data('w1', [24, 50], append_batch_size=False)
        w2 = fluid.layers.data('w2', [50, 14], append_batch_size=False)
        yg = fluid.layers.data('yg', [2, ], append_batch_size=False)

        inp.stop_gradient = False
        w1.stop_gradient = False
        w2.stop_gradient = False
        yg.stop_gradient = False

        y = fluid.layers.mul(fluid.layers.mul(inp, w1), w2)

        f = fluid.layers.reshape(y, [-1])

        x = [w1, w2]
        dfdx_f1 = fluid.gradients(f, x, f)

        # double gradient
        dfdx_x_ = fluid.gradients([dfdx_f1[0]*x[0], dfdx_f1[1]*x[1]], f)
        print(dfdx_x_)
        print( train_program, file=open("program.txt", 'wt'))
    exe = fluid.Executor(place)
    exe.run(program=fluid.default_startup_program())
    compiled_prog = fluid.compiler.CompiledProgram(train_program)

    res = exe.run(compiled_prog, feed={'inp':inp_, 'w1':w1_, 'w2':w2_, 'yg':yg_},
                  fetch_list=[dfdx_f1[0].name, dfdx_f1[1].name,
                              dfdx_x_[0].name])
    print(res)
test_grad_grad()
  
```

## 报错截图

image

计算第二次梯度的时候len(self.ops)和 len(ops_in_cpp)不相等；这段代码十月之前在paddle上还可以正常运行

### PR DESCRIPTION
**Pull request by ceci3**
fix https://github.com/PaddlePaddle/Paddle/issues/20790
if targets of gradients is a list,  1250 line in calc_gradient will invoke shape and fill_constant op twice, two fill_constant op after shape will be filtered by program, and append to the end of program once, it will cause the number of ops_in_python and ops_in_cpp is different.

### CODE REVIEW

File: python/paddle/fluid/backward.py [hunk 1]
```diff
@@ -1247,9 +1247,7 @@ def calc_gradient(targets, inputs, target_gradients=None, no_grad_set=None):
         target = targets[i]
         if grad is None:
             grad_name = _append_grad_suffix_(target.name)
-            target_shape = paddle.fluid.layers.shape(target)
-            op_desc = _create_op_desc_("fill_constant",
-                                       {"ShapeTensor": [target_shape.name]},
```
Comment (liym27) [inline python/paddle/fluid/backward.py:1252]: 之前加入输入`ShapeTensor`是为了保证传入的shape是正确的（不含-1）。
若删掉，fill_constant所使用的target.shape中可能有-1，将会导致底层内存分配出错。

**Comment by liym27** [comment]
之前加入输入`ShapeTensor`是为了保证传入的shape是正确的（不含-1）。
若删掉，fill_constant所使用的target.shape中可能有-1，将会导致底层内存分配出错。

File: python/paddle/fluid/backward.py [hunk 2]
```diff
@@ -1247,9 +1247,7 @@ def calc_gradient(targets, inputs, target_gradients=None, no_grad_set=None):
         target = targets[i]
         if grad is None:
             grad_name = _append_grad_suffix_(target.name)
-            target_shape = paddle.fluid.layers.shape(target)
```
Comment (liym27) [inline python/paddle/fluid/backward.py:1250]: 建议修改调用 `shape`接口的方式，通过函数_create_op_desc_()创建 `shape` op desc.

Comment (ceci3) [reply python/paddle/fluid/backward.py:1250]: done, thanks

File: python/paddle/fluid/backward.py (hunk 1 above)
Comment (ceci3) [reply python/paddle/fluid/backward.py:1252]: done, thanks

**Comment by liym27** [comment]
LGTM

**Review by qingqing01 — APPROVED** [review]

**Review by luotao1 — APPROVED** [review]
LGTM for backward.py

**Pull request merged**